### PR TITLE
[Fix] Generate Gmsh mesh in xz plane

### PR DESCRIPTION
**Pull request by TomFischer**
This fixes issue #663.

The original input data is rotated to the x-y-plane. All intermediate computations are done and finally to data is rotated back to the original location.

- [x] check with more test examples.

### CODE REVIEW

**Comment by waltherm** [comment]
looks good, works with my tested examples, which failed before. thanks, @TomFischer !

File: FileIO/GMSHInterface.cpp [hunk 1]
```diff
@@ -371,12 +374,13 @@ void GMSHInterface::writeGMSHInputFile(std::ostream& out)
...
 		return;
-	} else {
-		const std::size_t n_pnts(merged_pnts->size());
-		for (std::size_t k(0); k<n_pnts; k++) {
-			(*((*merged_pnts)[k]))[2] = 0.0;
-		}
 	}
+
+	// Rotate points to the x-y-plane.
+	MathLib::DenseMatrix<double> rot_mat = rotatePointsToXY(*merged_pnts);
```
Comment (endJunction) [inline FileIO/GMSHInterface.cpp]: Maybe this needs a translation to the origin too?

File: FileIO/GMSHInterface.cpp [hunk 2]
```diff
@@ -463,12 +467,54 @@ void GMSHInterface::writeGMSHInputFile(std::ostream& out)
...
-		if (_gmsh_pnts[k]) {
-			out << *(_gmsh_pnts[k]) << "\n";
+	for (auto & gmsh_pnt : _gmsh_pnts) {
+		// reverse rotation
+		if (gmsh_pnt) {
+			double* tmp = _inverse_rot_mat * gmsh_pnt->getCoords();
+			(*gmsh_pnt)[0] = tmp[0];
+			(*gmsh_pnt)[1] = tmp[1];
+			(*gmsh_pnt)[2] = tmp[2];
+			delete [] tmp;
```
Comment (rinkk) [inline FileIO/GMSHInterface.cpp]: why can you not simply assign ``` gmsh_pnt = _inverse_rot_mat * gmsh_pnt->getCoords();``` ?

File: FileIO/GMSHInterface.cpp [hunk 3]
```diff
@@ -463,12 +467,54 @@ void GMSHInterface::writeGMSHInputFile(std::ostream& out)
...
+			out << *gmsh_pnt << "\n";
 		}
 	}
 }
 
+MathLib::DenseMatrix<double> GMSHInterface::rotatePointsToXY(
```
Comment (rinkk) [inline FileIO/GMSHInterface.cpp]: i might be wrong but wasn't there a similar method somewhere else? you definitely needed sth like this in the TB example. if so, we should unify both methods...

**Comment by rinkk** [comment]
pls check for code duplication concerning rotation, otherwise :+1:

File: FileIO/GMSHInterface.cpp (hunk 1 above)
Comment (TomFischer) [inline FileIO/GMSHInterface.cpp]: This can be done. But I don't see a great benefit.

File: FileIO/GMSHInterface.cpp (hunk 2 above)
Comment (TomFischer) [inline FileIO/GMSHInterface.cpp]: The types do not match. The matrix multiplication is not overloaded for `GMSHPoint`s.

File: FileIO/GMSHInterface.cpp (hunk 3 above)
Comment (TomFischer) [inline FileIO/GMSHInterface.cpp]: Very good point. The only difference between `GMSHInterface::rotatePointsToXY()` and the `rotatePointsToXY()` from GeoLib/AnalyticalGeometry* is the return value. I'll try to change the return value of the GeoLib version and remove the code above.

File: FileIO/GMSHInterface.cpp [hunk 4]
```diff
@@ -371,12 +374,13 @@ void GMSHInterface::writeGMSHInputFile(std::ostream& out)
...
-		const std::size_t n_pnts(merged_pnts->size());
-		for (std::size_t k(0); k<n_pnts; k++) {
-			(*((*merged_pnts)[k]))[2] = 0.0;
-		}
 	}
+
+	// Rotate points to the x-y-plane.
+	MathLib::DenseMatrix<double> rot_mat = rotatePointsToXY(*merged_pnts);
+	// Compute inverse rotation matrix to reverse the rotation later on.
+	computeInverseRotationMatrix(rot_mat);
```
Comment (norihiro-w) [inline FileIO/GMSHInterface.cpp]: why not simply use a transpose of the rotation matrix? usually R^T = R^-1

Comment (TomFischer) [inline FileIO/GMSHInterface.cpp]: You are right! Thanks. The matrix has already a transpose operation.

**Comment by norihiro-w** [comment]
A basic question: why do you need to rotate points to xy? I thought Gmsh can handle points in xz

**Comment by TomFischer** [comment]
@norihiro-w 
> A basic question: why do you need to rotate points to xy? I thought Gmsh can handle points in xz 

You are right: Gmsh can handle input data located not only in the x-y-plane.

The pre-processing of the data (topological computations, intersections, ...) are performed under the assumption the data is located in the x-y-plane. For this reason the workflow implemented in this PR rotates the input data, perform these computations and afterwards rotates the data back (for this I need the inverse (transposed) rotation matrix).

**Comment by norihiro-w** [comment]
i see. thank you for the explanation!

**Pull request merged**